src/vote_graph.rs

```diff
@@ -0,0 +1,515 @@
+	ancestors: Vec<H>,
+	descendents: Vec<H>, // descendent vote-nodes
+	cumulative_vote: V,
+}
+
+impl<H: Hash + PartialEq + Clone, V> Entry<H, V> {
+	// whether the given hash, number pair is a direct ancestor of this node.
+	// `None` signifies that the graph must be traversed further back.
+	fn in_direct_ancestry(&self, hash: &H, number: usize) -> Option<bool> {
+		self.ancestor_block(number).map(|h| h == hash)
+	}
+
+	// Get ancestor block by number. Returns `None` if there is no block
+	// by that number in the direct ancestry.
+	fn ancestor_block(&self, number: usize) -> Option<&H> {
+		if number >= self.number { return None }
+		let offset = self.number - number - 1;
+
+		self.ancestors.get(offset)
+	}
+
+	// get ancestor vote-node.
+	fn ancestor_node(&self) -> Option<H> {
+		self.ancestors.last().map(|x| x.clone())
+	}
+}
+
+/// Maintains a DAG of blocks in the chain which have votes attached to them,
+/// and vote data which is accumulated along edges.
+pub struct VoteGraph<H: Hash + Eq, V> {
+	entries: HashMap<H, Entry<H, V>>,
+	heads: HashSet<H>,
+	base: H,
+}
+
+impl<H, V> VoteGraph<H, V> where
+	H: Hash + Eq + Clone + Ord,
+	V: ::std::ops::AddAssign + Default + Clone,
+{
+	fn new(base_hash: H, base_number: usize) -> Self {
+		let mut entries = HashMap::new();
+		entries.insert(base_hash.clone(), Entry {
+			number: base_number,
+			ancestors: Vec::new(),
+			descendents: Vec::new(),
+			cumulative_vote: V::default(),
+		});
+
+		let mut heads = HashSet::new();
+		heads.insert(base_hash.clone());
+
+		VoteGraph {
+			entries,
+			heads,
+			base: base_hash,
+		}
+	}
+
+	/// Insert a vote with given value into the graph at given hash and number.
+	pub fn insert<C: Chain<H>>(&mut self, hash: H, number: usize, vote: V, chain: &C) -> Result<(), Error> {
+		match self.find_containing_nodes(hash.clone(), number) {
+			Some(containing) => if containing.is_empty() {
+				self.append(hash.clone(), number, chain)?;
+			} else {
+				self.introduce_branch(containing, hash.clone(), number);
+			},
+			None => {}, // this entry already exists
+		}
+
+		// update cumulative vote data.
+		// NOTE: below this point, there always exists a node with the given hash and number.
+		let mut inspecting_hash = hash;
+		loop {
+			let active_entry = self.entries.get_mut(&inspecting_hash)
+				.expect("vote-node and its ancestry always exist after initial phase; qed");
+
+			active_entry.cumulative_vote += vote.clone();
+
+			match active_entry.ancestor_node() {
+				Some(parent) => { inspecting_hash = parent },
+				None => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	/// Find the best GHOST descendent of the given block.
+	/// Pass a closure used to evaluate the cumulative vote value.
+	///
+	/// The GHOST (hash, number) returned will be the block with highest number for which the
+	/// cumulative votes of descendents and itself causes the closure to evaluate to true.
+	///
+	/// This assumes that the evaluation closure is one which returns true for at most a single
```

Comment:
This holds assuming an honest majority will follow the heaviest chain.

Reply:
I think it holds as long as nobody double-votes (or in the case of supermajority, less than 1/3 double-vote). But I intend to filter out all double-votes in a higher level that instantiates the graph in a follow-up PR, so the property should always hold in our use-case.

I would like more unit tests to ensure that my logic is sound here